### PR TITLE
Clarify the use of JSON in requests and responses

### LINKED ISSUE #1182: Spec documentation does not say that request bodies are required
The OpenAPI definitions all have `required: true` for request bodies (meaning that clients must provide an empty JSON object even if they don't want to set any of the parameters), but this isn't reflected in the words of the spec.

### PR DESCRIPTION
**Pull request by richvdh**
Fixes #1182

Preview: https://pr1185--matrix-spec-previews.netlify.app

### CODE REVIEW

File: content/client-server-api/_index.md [hunk 1]
```diff
@@ -196,6 +208,8 @@ only read state (e.g.: `/sync`, get account data, etc).
 The user is unable to reject an invite to join the server notices room.
 See the [Server Notices](#server-notices) module for more information.
 
+### Transaction identifiers
```
Comment (richvdh) [inline content/client-server-api/_index.md:211]: Gave this a new heading to stop it being stuck under "Other error codes"

File: content/client-server-api/_index.md [hunk 2]
```diff
@@ -208,8 +222,6 @@ Some API endpoints may allow or require the use of `POST` requests
 without a transaction ID. Where this is optional, the use of a `PUT`
 request is strongly recommended.
 
-{{% http-api spec="client-server" api="versions" %}}
```
Comment (richvdh) [inline content/client-server-api/_index.md:211]: and moved this into "discovery" (for want of a better place), for the same reason.

File: content/identity-service-api.md [hunk 3]
```diff
@@ -38,8 +38,20 @@ Appendix.
 
 The mandatory baseline for identity server communication in Matrix is
 exchanging JSON objects over HTTP APIs. HTTPS is required for
-communication, and all API calls use a Content-Type of
-`application/json`. In addition, strings MUST be encoded as UTF-8.
+communication.
+
+All `POST` and `PUT` endpoints, with the exception of [`POST
+/_matrix/identity/v2/account/logout`](#post_matrixidentityv2accountlogout),
```
Comment (richvdh) [inline content/identity-service-api.md:44]: there's always an exception.

Comment (uhoreg) [reply content/identity-service-api.md:44]: Why is `/account/logout` an exception to this?

Comment (richvdh) [reply content/identity-service-api.md:44]: I suspect it's because nobody thought about it before, but: it just is, according to both the openapi definition, and the `sydent` implementation.

Comment (richvdh) [reply content/identity-service-api.md:44]: (and https://spec.matrix.org/v1.3/identity-service-api/#authentication, which says "no ... request body")

File: content/client-server-api/_index.md [hunk 4]
```diff
@@ -13,20 +13,32 @@ clients which maintain a full local persistent copy of server state.
...
+HTTPS is recommended for communication. The use of plain HTTP is not
+recommended outside test environments.
 
 Clients are authenticated using opaque `access_token` strings (see [Client
 Authentication](#client-authentication) for details).
 
+All `POST` and `PUT` endpoints, with the exception of
+[`POST /_matrix/media/v3/upload`](#post_matrixmediav3upload), require the
+client to supply a request body containing a (potentially empty) JSON object.
+Clients are *not* required to supply a `Content-Type` header.
```
Comment (uhoreg) [inline content/client-server-api/_index.md:28]: :( Can we at least say that clients "should" supply a `Content-Type: application/json` header?

Comment (richvdh) [reply content/client-server-api/_index.md:28]: I'm not really sure if that gains us much, unless we're planning to mandate it sometime soon? Servers still have to assume it's json even if there is no content-type header (or, for that matter, if the content-type header says it is something other than json, like text).

Comment (uhoreg) [reply content/client-server-api/_index.md:28]: To me, "Client are *not* required to supply a `Content-Type` header.` sounds like an encouragement to not bother sending a `Content-Type`, which just ... feels wrong to me.  At least if we say that they "should" send a `Content-Type`, we're encouraging people to send proper HTTP.  Maybe "Clients should supply an appropriate `Content-Type` header, but servers should be able to handle requests that do not have one, or an incorrect one."?  Though that's quite verbose. :shrug: This isn't a blocking comment; I'm just reluctant to encourage people to be sloppy.

File: content/identity-service-api.md (hunk 3 above)
Comment (uhoreg) [reply content/identity-service-api.md:44]: It might be nice to put a "for historical reasons" somewhere in that sentence (though I'm not sure where it would fit best) to indicate why it's like that, and that it isn't because that endpoint is actually special (as opposed to the other exceptions (media upload/download) which have actual reasons for not being JSON).  But if there's no nice way to fit that in, it's not the end of the world.

Comment (richvdh) [reply content/identity-service-api.md:44]:
```suggestion
All `POST` and `PUT` endpoints, with the exception (for historical reasons) of [`POST
/_matrix/identity/v2/account/logout`](#post_matrixidentityv2accountlogout),
```

File: content/client-server-api/_index.md (hunk 4 above)
Comment (richvdh) [reply content/client-server-api/_index.md:28]: I think "should" ~= "MUST", so just saying "should" implies that servers are allowed to enforce it.  So I could say this:

```suggestion
Clients should supply a `Content-Type` header of `application/json` for all requests with JSON bodies,
but this is not required.
```

but it feels a bit wordy. 

I don't really see what we gain, as guardians of the matrix spec, by discouraging people from this particular bit of sloppiness.

Comment (uhoreg) [reply content/client-server-api/_index.md:28]: > I don't really see what we gain, as guardians of the matrix spec, by discouraging people from this particular bit of sloppiness.

It feels to me that, since we're building on HTTP, we should be encouraging good HTTP behaviour, but maybe I'm being overly picky.

Your suggested wording seems fine to me, but if you don't think it's worth changing, then that's OK.

**Review by uhoreg — APPROVED** [review]

File: content/identity-service-api.md [hunk 5]
```diff
@@ -38,8 +38,20 @@ Appendix.
...
 The mandatory baseline for identity server communication in Matrix is
 exchanging JSON objects over HTTP APIs. HTTPS is required for
-communication, and all API calls use a Content-Type of
-`application/json`. In addition, strings MUST be encoded as UTF-8.
+communication.
+
+All `POST` and `PUT` endpoints, with the exception (for historical reasons) of [`POST
+/_matrix/identity/v2/account/logout`](#post_matrixidentityv2accountlogout),
+require the client to supply a request body containing a (potentially empty)
+JSON object. Clients are *not* required to supply a `Content-Type` header.
```
Comment (richvdh) [inline content/identity-service-api.md:46]:
```suggestion
JSON object. Clients should supply a `Content-Type` header of `application/json`
for all requests with JSON bodies, but this is not required.
```

File: content/server-server-api.md [hunk 6]
```diff
@@ -47,13 +47,41 @@ an HTTPS PUT request.
...
-low-bandwidth/low-roundtrip mobile usage. For the default HTTP
-transport, all API calls use a Content-Type of `application/json`. In
-addition, all strings MUST be encoded as UTF-8.
+The mandatory baseline for server-server communication in Matrix is
+exchanging JSON objects over HTTPS APIs. More efficient transports may be
+specified in future as optional extensions.
+
+All `POST` and `PUT` endpoints require the requesting server to supply a
+request body containing a (potentially empty) JSON object. Requesting servers
+are *not* required to supply a `Content-Type` header.
```
Comment (richvdh) [inline content/server-server-api.md:56]:
```suggestion
request body containing a (potentially empty) JSON object. Requesting servers
should supply a `Content-Type` header of `application/json` for all requests
with JSON bodies, but this is not required.
```

File: content/identity-service-api.md [hunk 7]
```diff
@@ -38,8 +38,20 @@ Appendix.
...
+
+Similarly, all endpoints require the server to return a JSON object.  Servers
+should include a `Content-Type` header of `application/json` for all JSON
+responses.
```
Comment (richvdh) [inline content/identity-service-api.md:50]:
```suggestion
Similarly, all endpoints require the server to return a JSON object.  Servers
must include a `Content-Type` header of `application/json` for all JSON
responses.
```

File: content/client-server-api/_index.md [hunk 8]
```diff
@@ -13,20 +13,33 @@ clients which maintain a full local persistent copy of server state.
...
+[`POST /_matrix/media/v3/upload`](#post_matrixmediav3upload), require the
+client to supply a request body containing a (potentially empty) JSON object.
+Clients should supply a `Content-Type` header of `application/json` for all requests with JSON bodies,
+but this is not required.
+
+Similarly, all endpoints require the server to return a JSON object,
+with the exception of 200 responses to
+[`GET /_matrix/media/v3/download/{serverName}/{mediaId}`](#get_matrixmediav3downloadservernamemediaid)
+and [`GET /_matrix/media/v3/thumbnail/{serverName}/{mediaId}`](#get_matrixmediav3thumbnailservernamemediaid).
+Servers should include a `Content-Type` header of `application/json` for all JSON responses.
```
Comment (richvdh) [inline content/client-server-api/_index.md:35]:
```suggestion
Servers msut include a `Content-Type` header of `application/json` for all JSON responses.
```

File: content/server-server-api.md [hunk 9]
```diff
@@ -47,13 +47,41 @@ an HTTPS PUT request.
...
+
+Similarly, all endpoints in this specification require the destination server
+to return a JSON object.  Servers should include a `Content-Type` header of
```
Comment (richvdh) [inline content/server-server-api.md:59]:
```suggestion
to return a JSON object.  Servers must include a `Content-Type` header of
```

File: content/client-server-api/_index.md [hunk 10]
```diff
@@ -13,20 +13,33 @@ clients which maintain a full local persistent copy of server state.
...
+[`POST /_matrix/media/v3/upload`](#post_matrixmediav3upload), require the
+client to supply a request body containing a (potentially empty) JSON object.
+Clients should supply a `Content-Type` header of `application/json` for all requests with JSON bodies,
+but this is not required.
+
+Similarly, all endpoints require the server to return a JSON object,
+with the exception of 200 responses to
+[`GET /_matrix/media/v3/download/{serverName}/{mediaId}`](#get_matrixmediav3downloadservernamemediaid)
+and [`GET /_matrix/media/v3/thumbnail/{serverName}/{mediaId}`](#get_matrixmediav3thumbnailservernamemediaid).
+Servers msut include a `Content-Type` header of `application/json` for all JSON responses.
```
Comment (uhoreg) [inline content/client-server-api/_index.md:35]: s/msut/must/

Comment (richvdh) [reply content/client-server-api/_index.md:35]: fixed in https://github.com/matrix-org/matrix-spec/commit/3dc3150f6d8c1bf77bd41af2b7540c1632d18998
